internal/stolonctl/pkg/cluster/upgrade.go

```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
 		return trace.Wrap(err)
 	}
 
-	// TODO delete debugging info
-	log.Info(res)
 	if !crdclient.IsPhaseCompleted(res, crd.StolonUpgradePhaseInit) {
 		res, err = crdclient.MarkPhase(res, crd.StolonUpgradePhaseInit, crd.StolonUpgradeStatusCompleted)
 		if err != nil {
 			return trace.Wrap(err)
 		}
 	}
+	// Check status of cluster
+	res, err = upgradeControl.checkStatus(res, crd.StolonUpgradePhaseChecks)
+	if err != nil {
+		return trace.Wrap(err)
+	}
 
 	// Backup Postgres
-	if !crdclient.IsPhaseCompleted(res, crd.StolonUpgradePhaseBackupPostgres) {
-		res, err = crdclient.MarkPhase(res, crd.StolonUpgradePhaseBackupPostgres, crd.StolonUpgradeStatusInProgress)
-		if err != nil {
-			return trace.Wrap(err)
+	res, err = upgradeControl.executePhase(res, crd.StolonUpgradePhaseBackupPostgres, func() error {
+		err := upgradeControl.backupPostgres()
+		return err
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Delete keeper/sentinel daemonset/deployment
+	res, err = upgradeControl.executePhase(res, crd.StolonUpgradePhaseDeleteDeployment, func() error {
+		err := upgradeControl.deleteDeployment(stolonSentinelDeployment)
+		return err
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	res, err = upgradeControl.executePhase(res, crd.StolonUpgradePhaseDeleteDaemonset, func() error {
+		err := upgradeControl.deleteDaemonset(stolonKeeperDaemonset)
+		return err
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Create job to upgrade PostgreSQL schema on keeper master
+	res, err = upgradeControl.executePhase(res, crd.StolonUpgradePhaseUpgradePostgresSchema, func() error {
+		var masterNodeName string
+		standbyNodeNames := make(map[string]string)
+		for _, pod := range res.Spec.PodsStatus {
+			if pod.PodIP == res.Spec.ClusterData.KeepersState[res.Spec.ClusterData.ClusterView.Master].ListenAddress {
```

Comment:
I wonder - if you're referring to master so frequently - would it not be easier if `KeepersState` kept the master state separately? Which would also be mandatory - i.e. not a pointer